text/0010-modularity.md

```diff
@@ -0,0 +1,459 @@
+mechanisms that favour better composition, reuse and sharing of user
+defined logic and queries are needed.
+
+Currently the unit of modularity and unit of deployment in tremor are
+indivisible blobs of code - be they scripts or queries. This has worked
+very well to date but as the complexity and size of solutions built with
+tremor grows this is no longer tenable in the long term.
+
+Identified requirements
+
+-   As tremor has multiple DSLs, the building blocks for reusing units
+    of code from the file system should be common across
+    DSLs.
+
+-   For tremor-script, code should be modularisable through the
+    introduction of functions.
+
+-   For tremor-query, code should be modularisable through modular
+    definitions and/or reusable sub-queries
+
+-   Modules should be nestable on the file system, and within DSLs with
+    a consistent means to reference units of code or values defined
+    within nested modules regardless of their origin ( within the same
+    unit of code, from some external module ).
+
+-   There should be a mechanism to load external libraries or packages
+    of code from well-known locations.
+
+-   The module mechanism should be usable by multiple DSLs with minimal
+    effort and with the same basic behaviour, structure. However, this
+    RFC does not place constraints on any DSL specifics per se.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Elements of modular user defined logic in the tremor project.
+
+## Module Path
+[module_path]: #module_path
+
+A module path is a set of URLs ( normatively directories on a file
+system ) that form the root of a set of related modules.
+
+On Linux/Unix module paths are provided via the TREMOR_PATH environment
+variable and they are separated by ':' ( colon ). Paths that are not
+readable or that do not exist are ignored.
+
+Example path:
+
+```bash
+TREMOR_PATH="/etc/tremor/lib:/opt/shared/framework/lib:/opt/myproject/mylib"
+```
+
+## Modules
+[modules]: #modules
+
+Modules in tremor are the lowest unit of compilation available to
+developers to modularise tremor logic across multiple logical
+namespaces. On the filesystem, modules are rooted at a base path and are
+nested with folders. Within a file nesting is via the mod
+clause.
+
+In tremor-script, only the top-level module can capture events or or
+mutate state. Modules loaded via the module system are restricted to
+const, fn, and intrinsic expressions. By design constraint at this time,
+tremor-script is biased towards pure side-effect free functional
+programming.
+
+In tremor-query, only the top-level module can create nodes in the
+active query pipeline graph. A module logically encapsulates a reusable
+sub-graph in a query pipeline. The definitions of windows, operators or
+scripts can be reused. Within embedded scripts, modules used in scripts
+are constrained to the rules for modules for tremor-script.
```

Comment:
trickle modules can also include functions and constants - this allows using script modules in trickle to for example have functions available in group by, having, of where in select

Reply:
👍 Yep, that's worth an explicit mention as the rfc leaves it as implied whereas its explicitly supported.